EDGE3D App

pguthrey opened the pull request:
# The EDGE3D kernel computes the summed mass + stiffness matrix of low order edge bases, which is relevant to MHD discretizations.

- This PR is a feature
- It does the following (modify list as needed):
  - Adds the EDGE3D app

artv3 commented on src/apps/EDGE3D.cpp:
```diff
@@ -0,0 +1,125 @@
+  setFLOPsPerRep(number_of_elements * flops_per_element);
+
+  // No idea what this is?
+  checksum_scale_factor = 0.001 *
```
I believe this is used to scale terms in the output array in a manner that helps avoid getting the right answer by accident. I'm forgetting the details, but perhaps @rhornung67  can help explain a bit.

rhornung67 replied:
Depending on a kernel's properties (default problem size, default # samples, how a checksum is calculated, typical magnitude of result values used to calculate checksums, etc.) we sometimes ad a scale factor to the checksum calculation to avoid overflow and accuracy issues. Typically, we try to shoot for checksum differences < 1e-8 between variants for correct results.

artv3 commented on src/apps/EDGE3D.hpp:
```diff
@@ -0,0 +1,527 @@
+
+private:
+  //static const size_t default_gpu_block_size = 256;
+  //using gpu_block_sizes_type = gpu_block_size::make_list_type<default_gpu_block_size>;
```
Perhaps delete if not used?

pguthrey replied:
Any idea what this should be? I am inclined to uncomment this default value and delete what I used.

MrBurmark replied:
Is this resolved?

CRobeck commented on src/apps/EDGE3D-Cuda.cpp:
```diff
@@ -0,0 +1,129 @@
+                      Index_type ibegin, Index_type iend)
+{
+   Index_type ii = blockIdx.x * block_size + threadIdx.x;
+   Index_type i = ii + ibegin;
```
Why do we need two indexes here? `ii` appears to be unused.

pguthrey replied:
i = ii + ibegin ?

MrBurmark replied:
I think the suggestion here is `Index_type i = ibegin + blockIdx.x * block_size + threadIdx.x;`

MrBurmark commented on src/apps/EDGE3D.cpp:
```diff
@@ -0,0 +1,124 @@
+  setUsesFeature(Forall);
+
+  setVariantDefined( Base_Seq );
+  setVariantDefined( RAJA_Seq );
```
Where are the lambda variants?

pguthrey replied:
Fixed

MrBurmark commented on src/apps/EDGE3D.hpp:
```diff
@@ -0,0 +1,525 @@
+// SPDX-License-Identifier: (BSD-3-Clause)
+//~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~//
+/*
+const Real_type qpts_1d[NQ_1D] = { 1. };
```
where's the loop over zones?

MrBurmark commented on src/apps/EDGE3D.hpp:
```diff
@@ -0,0 +1,525 @@
+#define NQ_1D 2
+
+#define EDGE3D_BODY \
+  const int iz = i;\
```
do we need to rename i to iz?

pguthrey commented:
Todo: investigate launch bounds.

MrBurmark commented on src/apps/EDGE3D-Cuda.cpp:
```diff
@@ -0,0 +1,129 @@
+  deallocCudaDeviceData(sum);
+
+template < size_t block_size >
+__launch_bounds__(block_size)
```
Hmm, already have launch bounds.

artv3 commented on src/apps/EDGE3D-Hip.cpp:
```diff
@@ -0,0 +1,129 @@
+  deallocHipDeviceData(sum);
+
+template < size_t block_size >
+__launch_bounds__(block_size)
```
launch bounds is on the HIP on too.. weird..

MrBurmark commented on src/apps/EDGE3D.hpp:
```diff
@@ -0,0 +1,530 @@
+
+private:
+  static const size_t default_gpu_block_size = 256;
+  using gpu_block_sizes_type = gpu_block_size::list_type<default_gpu_block_size>;
```
```suggestion
  using gpu_block_sizes_type = gpu_block_size::make_list_type<default_gpu_block_size>;
```

MrBurmark commented on src/apps/EDGE3D.hpp:
```diff
@@ -0,0 +1,439 @@
+#include "ahelper.hpp"
+
+__host__ __device__
+inline void edge_MpSmatrix(
```
let's consider whether we think this should be in the macro or in a function

rhornung67 replied:
Also, it should be changed because it doesn't compile.

pguthrey replied:
Ah those pesky non-gpu builds... ;)

fixed

pguthrey commented on src/apps/ahelper.hpp:
```diff
@@ -0,0 +1,901 @@
+
+#pragma once
```
changed to "mixed_fem_helper"

rhornung67 commented:
@pguthrey I fixed all the issues in your PR and got all the checks to pass. I would like @MrBurmark to do a final review before merging.

MrBurmark commented on src/apps/EDGE3D.cpp:
```diff
@@ -0,0 +1,130 @@
+    matrix_size;
+
+  // touched data size, not actual number of stores and loads ?
+  setBytesPerRep( number_of_elements*reals_per_element*sizeof(Real_type) );
```
Is this just the size of loaded data, or loaded and stored data size?

MrBurmark replied:
Is this comment accurate?

pguthrey replied:
Let's go through the impl and determine that

MrBurmark commented on src/apps/EDGE3D.cpp:
```diff
@@ -0,0 +1,130 @@
+
+  setUsesFeature(Forall);
+
+  setVariantDefined( Base_Seq );
+  setVariantDefined( RAJA_Seq );
+  setVariantDefined( Lambda_Seq );
```
```suggestion
  setVariantDefined( Base_Seq );
  setVariantDefined( Lambda_Seq );
  setVariantDefined( RAJA_Seq );
```

MrBurmark commented on src/apps/EDGE3D.cpp:
```diff
@@ -0,0 +1,130 @@
+  setVariantDefined( Base_OpenMPTarget );
+  setVariantDefined( RAJA_OpenMPTarget );
+
+  setVariantDefined( Base_CUDA );
+  setVariantDefined( RAJA_CUDA );
+
+  setVariantDefined( Base_HIP );
+  setVariantDefined( RAJA_HIP );
```
what about lambda variants here?

MrBurmark commented on src/apps/mixed_fem_helper.hpp:
```diff
@@ -0,0 +1,939 @@
+// SPDX-License-Identifier: (BSD-3-Clause)
+//~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~//
+
+#pragma once
```
Shall we use ifdefs, I don't think we've used pragma once elsewhere in this code.

pguthrey commented on src/apps/EDGE3D.cpp:
```diff
@@ -0,0 +1,130 @@
+
+  setVariantDefined( Base_OpenMP );
+  setVariantDefined( RAJA_OpenMP );
+  setVariantDefined( Lambda_OpenMP );
```
Okay so now everything has lambda variants. I will monitor testing to see if it passes.

MrBurmark commented on src/apps/EDGE3D-Cuda.cpp:
```diff
@@ -0,0 +1,117 @@
+
+      RAJA::forall< RAJA::cuda_exec<block_size, true /*async*/> >(
+        RAJA::RangeSegment(ibegin, iend), [=] __device__ (Index_type i) {
+        edge3d_lam(i);
```
```suggestion
        EDGE3D_BODY;
```

MrBurmark commented on src/apps/EDGE3D-Cuda.cpp:
```diff
@@ -0,0 +1,117 @@
+
+  EDGE3D_DATA_SETUP;
+
+  auto edge3d_lam =
+    [=](Index_type i) {
+      EDGE3D_BODY;
+    };
+
```
```suggestion
```

MrBurmark commented on src/apps/EDGE3D-Hip.cpp:
```diff
@@ -0,0 +1,117 @@
+
+  EDGE3D_DATA_SETUP;
+
+  auto edge3d_lam =
+    [=](Index_type i) {
+      EDGE3D_BODY;
+    };
+
```
```suggestion
```

MrBurmark commented on src/apps/EDGE3D-Hip.cpp:
```diff
@@ -0,0 +1,117 @@
+
+      RAJA::forall< RAJA::hip_exec<block_size, true /*async*/> >(
+        RAJA::RangeSegment(ibegin, iend), [=] __device__ (Index_type i) {
+        edge3d_lam(i);
```
```suggestion
        EDGE3D_BODY;
```

MrBurmark reviewed (changes requested):
Extra lambdas in GPU variants

CRobeck commented on src/apps/EDGE3D.hpp:
```diff
@@ -0,0 +1,444 @@
+#include "mixed_fem_helper.hpp"
+
+RAJA_HOST_DEVICE
+inline void edge_MpSmatrix(
```
RAJA_INLINE

CRobeck commented on src/apps/EDGE3D.hpp:
```diff
@@ -0,0 +1,444 @@
+
+RAJA_HOST_DEVICE
+inline void edge_MpSmatrix(
+  const double  (&x)[NB],
```
double->Real_Type
int ->Index_Type

CRobeck commented on src/apps/mixed_fem_helper.hpp:
```diff
@@ -0,0 +1,942 @@
+
+RAJA_HOST_DEVICE
+constexpr double compute_detj(
+  const double jxx,
```
double->Real_Type

pguthrey replied:
Thanks for that @rhornung67

CRobeck commented on src/apps/mixed_fem_helper.hpp:
```diff
@@ -0,0 +1,942 @@
+  }
+}
+
+template<int M, int P>
```
Would this make more sense as constexpr instead of a non type template parameter?

pguthrey replied:
Resolved in person

CRobeck commented on src/apps/mixed_fem_helper.hpp:
```diff
@@ -0,0 +1,942 @@
+    jxx*jyz*jzy - jxy*jyx*jzz + jxx*jyy*jzz;
+}
+
+template<int M>
```
Would this make more sense as constexpr instead of a non type template parameter?

pguthrey replied:
Resolved in person

CRobeck commented on src/apps/mixed_fem_helper.hpp:
```diff
@@ -0,0 +1,942 @@
+  const double  xyloc,
+  const double  tmpxyloc)
+{
+  (void) x[0];
```
1. Is the cast to void just for compiler warning? 
2. Can we remove the y and z from this scope block?
3. Is the bottom section of this scope block unreachable?

pguthrey replied:
1.  Yes
3.  The return statement is the last line that is called.

pguthrey replied:
Done

pguthrey commented on src/apps/EDGE3D.cpp:
```diff
@@ -0,0 +1,131 @@
+
+  // touched data size, not actual number of stores and loads
+  // see VOL3D.cpp
+  // 1
```
```suggestion
```

rhornung67 commented:
@MrBurmark @pguthrey I merged develop into this, fixed build issues, and addressed several issues in reviewer comments. Is it good to merge?

pguthrey commented:
> @MrBurmark @pguthrey I merged develop into this, fixed build issues, and addressed several issues in reviewer comments. Is it good to merge?

My plan is to finish this MR during the hackathon this week. I'll let you know once all the interested parties are happy.

pguthrey commented:
Ah I see the spelling was fixed in
https://github.com/LLNL/RAJAPerf/commit/0987b08e1695b9a637fa8086839ab93b7fda8c96

MrBurmark commented on src/apps/EDGE3D-Cuda.cpp:
```diff
@@ -0,0 +1,112 @@
+
+      const size_t grid_size = RAJA_DIVIDE_CEILING_INT(iend, block_size);
+
+      edge3d<block_size><<<grid_size, block_size>>>(sum,
```
Need to add stream/resource use to all of the cuda/hip kernels, see the other cuda/hip kernels for an example.

pguthrey replied:
Done

pguthrey commented on src/apps/mixed_fem_helper.hpp:
```diff
@@ -0,0 +1,942 @@
+//
+
+RAJA_HOST_DEVICE
+inline void LinAlg_qrule_Lobatto(
```
Done

CRobeck commented on src/apps/EDGE3D-OMP.cpp:
```diff
@@ -0,0 +1,100 @@
+    case Lambda_OpenMP : {
+
+      startTimer();
+      for (RepIndex_type irep = 0; irep < run_reps; ++irep) {
```
Can we be consistent between RepIndex_type and Index_type in these loops?

rhornung67 replied:
Thanks for catching that @CRobeck

MrBurmark replied:
Did you add an issue for this?

pguthrey commented:
Something wrong with Lambda_HIP. Will investigate.

MrBurmark commented on src/apps/EDGE3D-Cuda.cpp:
```diff
@@ -0,0 +1,118 @@
+      auto edge3d_lam = [=] __device__ (Index_type i) { EDGE3D_BODY; };
+
+      lambda_cuda_forall<block_size><<<grid_size, block_size, shmem, res.get_stream()>>>(
+        0, iend, edge3d_lam);
```
```suggestion
        ibegin, iend, edge3d_lam);
```

MrBurmark commented on src/apps/EDGE3D.cpp:
```diff
@@ -0,0 +1,132 @@
+
+  // touched data size, not actual number of stores and loads
+  // see VOL3D.cpp
+  size_t reads_per_node = 3*sizeof(Real_type);
+  size_t writes_per_rep = 1*sizeof(Real_type);
+  setBytesPerRep( writes_per_rep * getItsPerRep() +
+                  reads_per_node * (getItsPerRep() + 1+m_domain->jp+m_domain->kp) );
```
I like the idea of using named variables here, big improvement in readability. I'll add an issue to refactor this for the rest of the perf suite.
```suggestion
  size_t reads_per_node = 3*sizeof(Real_type);
  size_t writes_per_zone = 1*sizeof(Real_type);
  setBytesPerRep( writes_per_zone * getItsPerRep() +
                  reads_per_node * (getItsPerRep() + 1+m_domain->jp+m_domain->kp) );
```

pguthrey commented:
@MrBurmark I think this is good to go.

rhornung67 reviewed (approved):
We can merge this when checks come back all green.

rhornung67 commented:
@MrBurmark @pguthrey checks are all green. Merge this?
